pvlib/spectrum/mismatch.py

```diff
@@ -0,0 +1,232 @@
+        One or more measured solar irradiance spectra in a pandas.DataFrame
+        having wavelength in nm as column index.  A single spectrum may be
+        be given as a pandas.Series having wavelength in nm as index.
+        [(W/m^2)/nm]
+
+    e_ref: pandas.Series, optional
+        The reference spectrum to use for the mismatch calculation.
+        The index of the Series must contain wavelength values in nm.
+        The default is the ASTM G173-03 global tilted spectrum. [(W/m^2)/nm]
+
+    Returns
+    -------
+    smm: pandas.Series or float if a single measured spectrum is provided. [-]
+
+    Notes
+    -----
+    Measured solar spectral irradiance usually covers a wavelength range
+    that is smaller than the range considered as broadband irradiance.
+    The infrared limit for the former typically lies around 1100 or 1600 nm,
+    whereas the latter extends to around 2800 or 4000 nm.  To avoid a mismatch
+    between the magnitudes of the integrated spectra (the broadband values)
+    this function truncates the reference spectrum to the same range as the
+    measured (or simulated) field spectra.
+
+    If the default reference spectrum is used it is linearly interpolated
+    to the wavelengths of the measured spectrum, but if a reference spectrum
+    is provided via the parameter ``e_ref`` it is used without change. This
+    makes it possible to avoid interpolation, or to use a different method of
+    interpolation, or to avoid truncation.
+
+    The spectral response is linearly interpolated to the wavelengths of each
+    spectrum with which is it multiplied internally (``e_sun`` and ``e_ref``).
+    If the wavelengths of the spectral response already match one or both
+    of these spectra interpolation has no effect; therefore, another type of
+    interpolation could be used to process ``sr`` before calling this function.
+
+    The standards describing mismatch calculations focus on indoor laboratory
+    applications, but are applicable to outdoor performance as well.
+    The 2016 version of ASTM E973 [1]_ is somewhat more difficult to
+    read than the 2010 version [2]_ because it includes adjustments for
+    the temperature dependency of spectral response, which led to a
+    formulation using quantum efficiency (QE).
+    IEC 60904-7 is clearer and also discusses the use of a broadband
+    reference device. [3]_
+
+    References
+    ----------
+    .. [1] ASTM "E973-16 Standard Test Method for Determination of the
+       Spectral Mismatch Parameter Between a Photovoltaic Device and a
+       Photovoltaic Reference Cell" :doi:`10.1520/E0973-16R20`
+    .. [2] ASTM "E973-10 Standard Test Method for Determination of the
+       Spectral Mismatch Parameter Between a Photovoltaic Device and a
+       Photovoltaic Reference Cell" :doi:`10.1520/E0973-10`
+    .. [3] IEC 60904-7 "Computation of the spectral mismatch correction
+       for measurements of photovoltaic devices"
+    """
```

Comment:
What about IEC 61853-3 (ed. 1.0), specifically, equation (6) in section 7.3?

This is where the implementation here does not cover that formula precisely.

Reply:
Could you elaborate a little bit on where you see the problem?

Reply:
In any case, the equation shows the pre-integrated values of the reference and sun/field spectra in a similar manner to your proposed function signature. Do the pre-integrated values need to be different from the values that would be obtained from an actual integration?

I have some comments about the problems with this calculation of the standard in the docstring of my implementation in [pvpltools](https://github.com/adriesse/pvpltools-python/blob/master/pvpltools/iec61853.py#L127).

Reply:
@adriesse Sorry that I have not been able to look at this closely again this week.

> Do the pre-integrated values need to be different from the values that would be obtained from an actual integration?

This is perhaps the key question, assuming that the IEC 61853-3 use case is an important one for this code addition. (I have some issues with this formula, but it is what it is.) At this point I am assuming that this use case is important, and it requires some consideration in addition to the "primary reference cell calibration with a cavity radiometer" use case that I previously examined here. I intend to look closer at this this weekend.

There is also one last use case, which I am personally interested in. This is for PVfit's "irradiance" calculation of `F=Isc/Isc0`. The spectral-mismatch approach here provides an alternative to the approach that I presented at the recent PVPMC workshop. While I would like to see this code cover this case to my satisfaction, please know that I am not trying to hold up this PR based on this criterion.

Reply:
I might prefer to see future functions for specific use cases such as calibration or energy rating.  A single generic function could be fewer lines of code, but may not be as inviting for users who are not as broadly informed on the subject.

I have been involved in many discussions about IEC 61853-3.  The next version will look different, probably skipping the mismatch or spectral correction factor in this form altogether.  Anyway, it would be hard to cover the current version in this function due to the use of banded solar spectral irradiance data.

Reply:
I derived equation (6) in section 7.3 in IEC 61853-3 (ed. 1.0) from first principles. Recall that this is used in IEC 61853-1 is to compute the equivalent irradiance under the standard spectrum that would generate the same short-circuit current in the device under the given operating condition, denote here as `E_equiv`. One issue I found is that technically the spectral response of the device should depend on temperature. A second issue to note is that the upper limit of integration on the integrals, 'lambda_e", is fixed at 4 microns, which corresponds to the assumed spectral-response cutoff of total POA irradiance as measured by a pyranometer and (presumably) reported in the profiles in IEC 61853-4. In the study below, a cutoff change to 1.7 microns requires scaling the total POA irradiance accordingly.

I then computed the SSM and `E_equiv` that would be computed by code under review vs. an implementation that covers equation (6) in IEC 61853-3. The error appears small for the 4 micron cutoff, but becomes very significant for the 1.7 micron cutoff. What is quite interesting, and what I did not expect, is that the SMM changes significantly based on the wavelength cutoff, and this change is necessary to compute a consistent `E_equiv`! Of course, these results are contingent upon 3rd-party review :).

This is the modified source code to run the comparison. It can be used to compute both algorithms.
```python
def get_am15g(wavelength=None):
    # Contributed by Anton Driesse (@adriesse), PV Performance Labs. Aug. 2022
    # Modified by Mark Campanelli for testing purposes.

    pvlib_path = pvlib.__path__[0]
    filepath = os.path.join(pvlib_path, 'data', 'astm_g173_am15g.csv')

    am15g = pd.read_csv(filepath, index_col=0).squeeze()  # 0.28 to 4 microns
    am15g_integral_to_4_micron = np.trapz(am15g, x=am15g.index, axis=-1)  # W/m^2
    iec_integral_to_4_micron = 997.47  # W/m^2, from IEC 60904-3 Ed. 2 (2008)
    am15g_scaled_to_iec = iec_integral_to_4_micron / am15g_integral_to_4_micron * am15g

    if wavelength is not None:
        interpolator = interp1d(am15g.index, am15g,
                                kind='linear',
                                bounds_error=False,
                                fill_value=0.0,
                                copy=False,
                                assume_sorted=True)

        am15g = pd.Series(data=interpolator(wavelength), index=wavelength)

        interpolator = interp1d(am15g_scaled_to_iec.index, am15g_scaled_to_iec,
                                kind='linear',
                                bounds_error=False,
                                fill_value=0.0,
                                copy=False,
                                assume_sorted=True)

        am15g_scaled_to_iec = pd.Series(data=interpolator(wavelength), index=wavelength)

    am15g.index.name = 'wavelength'
    am15g.name = 'am15g'

    am15g_scaled_to_iec.index.name = 'wavelength'
    am15g_scaled_to_iec.name = 'am15g'

    return am15g, am15g_scaled_to_iec


def calc_spectral_mismatch_field(sr, e_sun, e_sun_tot, e_ref=None, e_ref_tot=None):
    # Contributed by Anton Driesse (@adriesse), PV Performance Labs. Aug. 2022
    # Modified by Mark Campanelli for testing purposes.

    # get the reference spectrum at wavelengths matching the measured spectra
    if e_ref is None:
        e_ref, _ = get_am15g(wavelength=e_sun.T.index)

    # a helper function to make usable fraction calculations more readable
    def integrate(e):
        return np.trapz(e, x=e.T.index, axis=-1)

    if e_sun_tot is None:
        # Use integrated spectral irradiance, typically an approximation
        # that misses "tail" of distribution.
        e_sun_tot = integrate(e_sun)

    if e_ref_tot is None:
        # Use integrated spectral irradiance, typically an approximation
        # that misses "tail" of distribution.
        e_ref_tot = integrate(e_ref)

    # interpolate the sr at the wavelengths of the spectra
    # reference spectrum wavelengths may differ if e_ref is from caller
    sr_sun = np.interp(e_sun.T.index, sr.index, sr, left=0.0, right=0.0)
    sr_ref = np.interp(e_ref.T.index, sr.index, sr, left=0.0, right=0.0)

    # calculate usable fractions
    uf_sun = integrate(e_sun * sr_sun) / e_sun_tot
    uf_ref = integrate(e_ref * sr_ref) / e_ref_tot

    # mismatch is the ratio or quotient of the usable fractions
    smm = uf_sun / uf_ref

    if isinstance(e_sun, pd.DataFrame):
        smm = pd.Series(smm, index=e_sun.index)

    return smm
```

This is the Jupyter notebook code that runs the comparison:
```python
import numpy as np
from pvlib import spectrum
from pvlib.tests import test_spectrum

# Standard spectra from 0.28 to 4 micron, e_ref_iec is am15g (from ASTM G173-03) scaled to "match" IEC 60904-3 Ed. 2 (2008).
e_ref_am15g_astm, e_ref_am15g_iec = spectrum.get_am15g()

# IEC 60904-3 Ed. 2 (2008), cumulatives are provided in standard.
e_ref_am15g_iec_idx_to_1p7_micron = e_ref_am15g_iec.T.index <= 1700
e_ref_am15g_iec_integral_to_1p7_micron = 942.88  # W/m^2
e_ref_am15g_iec_integral_to_4_micron = 997.47  # W/m^2
e_ref_am15g_iec_integral_to_infinity = 1000.  # W/m^2

# ASTM G173-03, calculate cumulatives not provided in standard.
e_ref_am15g_astm_idx_to_1p7_micron = e_ref_am15g_astm.T.index <= 1700
e_ref_am15g_astm_integral_to_1p7_micron = np.trapz(e_ref_am15g_astm[e_ref_am15g_astm_idx_to_1p7_micron], x=e_ref_am15g_astm.index[e_ref_am15g_astm_idx_to_1p7_micron], axis=-1)  # W/m^2
e_ref_am15g_astm_integral_to_4_micron = np.trapz(e_ref_am15g_astm, x=e_ref_am15g_astm.index, axis=-1)  # W/m^2
e_ref_am15g_astm_integral_to_infinity = None  # W/m^2, undefined in standard

# Get example solar spectrum from test fixture.
_, e_sun = test_spectrum.spectrl2_data.__pytest_wrapped__.obj()  # 0.3 to 4 micron
e_sun = e_sun.set_index('wavelength')
e_sun = e_sun.transpose().loc['specglo']
e_sun_idx_to_1p7_micron = e_sun.T.index <= 1700
e_sun_integral_to_1p7_micron = np.trapz(e_sun[e_sun_idx_to_1p7_micron], x=e_sun.T.index[e_sun_idx_to_1p7_micron], axis=-1)
e_sun_integral_to_4_micron = np.trapz(e_sun, x=e_sun.T.index, axis=-1)  # 708.4855034837434 W/m^2
scale_factor_estimate = e_ref_am15g_iec_integral_to_infinity / e_ref_am15g_iec_integral_to_4_micron
e_sun_integral_to_infinity = scale_factor_estimate * e_sun_integral_to_4_micron

# Relative spectral response for Si, 0.29 to 1.19 micron.
sr = spectrum.get_example_spectral_response()

# SMM computations IEC 61853-1 imply reference temperature in all situations, which is technically incorrect.

# Calculate SMM using formula in IEC 61853-1, using IEC 60904-3 Ed. 2 (2008) spectrum.
# Different cutoff wavelengths used for solar total irradiance integral according to the spectraradiometer cutoff.
iec_smm_inf = spectrum.calc_spectral_mismatch_field(
    sr, e_sun, e_sun_integral_to_infinity, e_ref=e_ref_am15g_iec, e_ref_tot=e_ref_am15g_iec_integral_to_infinity)
iec_smm_4_micron = spectrum.calc_spectral_mismatch_field(
    sr, e_sun, e_sun_integral_to_4_micron, e_ref=e_ref_am15g_iec, e_ref_tot=e_ref_am15g_iec_integral_to_infinity)
iec_smm_1p7_micron = spectrum.calc_spectral_mismatch_field(
    sr, e_sun[e_sun_idx_to_1p7_micron], e_sun_integral_to_1p7_micron, e_ref=e_ref_am15g_iec, e_ref_tot=e_ref_am15g_iec_integral_to_infinity)

# Calculate SMM using proposed formula double truncation of integrals with ASTM G173-03 spectrum.
astm_smm_4_micron = spectrum.calc_spectral_mismatch_field(sr, e_sun, None)
astm_smm_1p7_micron = spectrum.calc_spectral_mismatch_field(sr, e_sun[e_sun_idx_to_1p7_micron], None)

print("IEC am15g totals to 1.7 um, 4 um, and infinity (W/m^2):", e_ref_am15g_iec_integral_to_1p7_micron, e_ref_am15g_iec_integral_to_4_micron, e_ref_am15g_iec_integral_to_infinity)
print("ASTM am15g totals to 1.7 um, 4 um, and infinity (W/m^2):", e_ref_am15g_astm_integral_to_1p7_micron, e_ref_am15g_astm_integral_to_4_micron, e_ref_am15g_astm_integral_to_infinity)
print("e_sun totals to 1.7 um, 4 um, and infinity (W/m^2):", e_sun_integral_to_1p7_micron, e_sun_integral_to_4_micron, e_sun_integral_to_infinity)

print("IEC smm to infinity:", iec_smm_inf)
print("IEC smm to 4 um:", iec_smm_4_micron)
print("IEC smm to 1.7 um:", iec_smm_1p7_micron)
print("IEC am15g-equivalent E to infinity:", iec_smm_inf*e_sun_integral_to_infinity)
print("IEC am15g-equivalent E to 4 um:", iec_smm_4_micron*e_sun_integral_to_4_micron)
print("IEC am15g-equivalent E to 1.7 um:", iec_smm_1p7_micron*e_sun_integral_to_1p7_micron)

print("ASTM smm to 4 um:", astm_smm_4_micron)
print("ASTM smm to 1.7 um:", astm_smm_1p7_micron)
print("ASTM am15g-equivalent E to 4 um:", astm_smm_4_micron*e_sun_integral_to_4_micron)
print("ASTM am15g-equivalent E to 1.7 um:", astm_smm_1p7_micron*e_sun_integral_to_1p7_micron)
```

Here is the output for the comparison. Note the consistency of `E_equiv` in the "IEC version" of the computation.
```terminal
IEC am15g totals to 1.7 um, 4 um, and infinity (W/m^2): 942.88 997.47 1000.0
ASTM am15g totals to 1.7 um, 4 um, and infinity (W/m^2): 945.6188921247738 1000.3706555734423 None
e_sun totals to 1.7 um, 4 um, and infinity (W/m^2): 668.028852782929 708.4855034837434 710.2825182549283
IEC smm to infinity: 0.9913868433999364
IEC smm to 4 um: 0.9939014139772988
IEC smm to 1.7 um: 1.054093308636961
IEC am15g-equivalent E to infinity: 704.1647436949111
IEC am15g-equivalent E to 4 um: 704.1647436949111
IEC am15g-equivalent E to 1.7 um: 704.164743694911
ASTM smm to 4 um: 0.9923965546943565
ASTM smm to 1.7 um: 0.9902678450884945
ASTM am15g-equivalent E to 4 um: 703.0985727081635
ASTM am15g-equivalent E to 1.7 um: 661.5274925022902
```

Reply:
Thanks again for your detailed analysis, Mark.  I agree that there is potential for small differences with the tails, and large differences due to the cut-off at 1700 nm.  In this field application the cutoff at 1700 nm is only made if there is no information available about the spectrum beyond.  In that situation there are not a lot of choices available, and cutting off the reference spectrum at the same wavelength appears to be a reasonable solution.